syntax = "proto3";

package myfirstgrpc;

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}
  // Server streaming
  rpc StreamGreetings (HelloRequest) returns (stream HelloReply) {}
  // Client streaming
  rpc SendGreetings (stream HelloRequest) returns (HelloReply) {}
  // Bidirectional streaming
  rpc Chat (stream HelloRequest) returns (stream HelloReply) {}
  // File download
  rpc DownloadFile (FileRequest) returns (stream FileChunk) {}
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}

// The request message for file download.
message FileRequest {
  string filename = 1;
}

// The response message containing a chunk of the file.
message FileChunk {
  bytes content = 1;
}